pangeo_forge_recipes/patterns.py

```diff
@@ -267,10 +268,9 @@ def prune_pattern(fp: FilePattern, nkeep: int = 2) -> FilePattern:
         else:  # pragma: no cover
             assert "Should never happen"
 
-    return FilePattern(
-        fp.format_function,
-        *new_combine_dims,
-        fsspec_open_kwargs=fp.fsspec_open_kwargs,
-        query_string_secrets=fp.query_string_secrets,
-        is_opendap=fp.is_opendap,
-    )
+    sig = inspect.signature(fp.__init__)
+    kwargs = {
+        param: getattr(fp, param) for param in sig.parameters.keys()
+        if param not in ['format_function', 'combine_dims']
+    }
```

Comment:
👌

Reply:
Ok phew 😅  wasn't sure if this was getting a little too tricky, glad to get your 👌 

I agree having to change the prune function every time the signature of FilePattern.__init__ changes is asking for problems... but as a follow-on to our discussion in https://github.com/pangeo-forge/pangeo-forge-recipes/pull/167#discussion_r700633315, it looks like [dataclasses.replace](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace) doesn't work out-of-the-box for dataclasses with custom __init__s (which are required for variable numbers of *args):

> Be forewarned about how init=False fields work during a call to replace(). They are not copied from the source object, but rather are initialized in __post_init__(), if they’re initialized at all. It is expected that init=False fields will be rarely and judiciously used. If they are used, it might be wise to have alternate class constructors, or perhaps a custom replace() (or similarly named) method which handles instance copying.

So this was my workaround for that